### PR TITLE
Fix WriteTenants mutation

### PR DESCRIPTION
**Pull request by kaloyanspiridonov**
**Description**

Changes proposed in this pull request:
- Ensure that all tenant accesses are updated when the 'writeTenants' mutation is executed. When the 'writeTenants' or 'writeTenant' mutation is performed, any changes in the parent-child relationships are propagated correctly together with all corresponding tenant accesses.
- All four customer adoption metrics: "Customer Adoption", "Internal Customer Adoption", "External Customer Adoption" and "Trial Customer Adoption" have been fixed to display the correct status.
- When retrieving information about new, updated or deleted tenants from the external tenant repository, we occasionally encounter empty values for 'externalTenant', which breaks our tenant hierarchy. As a solution, we've implemented additional validation in the code to handle empty 'externalTenant' values and introduced a corresponding database constraint.

**Related issue(s)**

- ...

**Pull Request status**

- [x] Implementation
- [x] Unit tests
- [x] Integration tests
- [x] `chart/compass/values.yaml` is updated 
- [ ] Mocks are regenerated, using the automated script

### CODE REVIEW

File: chart/compass/charts/prometheus-postgres-exporter/values.yaml [hunk 1]
```diff
@@ -295,40 +295,40 @@ config:
...
       master: true
       cache_seconds: 43200
       metrics:
         - quantity:
             usage: "COUNTER"
-            description: "Number of TRIAL tenants that have at least one non auto-discovered Application (the tenant is owner and its parent is null) or tenants that have at least 1 formation different from the default one where their parent is either null or their parent is of type customer"
+            description: "Number of TRIAL tenants that have at least one non auto-discovered Application (the tenant is owner and its parent is null) or tenants that have at least on system in state different from 'INACTIVE' or tenants that have at least 1 formation different from the default one where their parent is either null or their parent is of type customer"
```
Comment (ivantenevvasilev) [inline chart/compass/charts/prometheus-postgres-exporter/values.yaml:331]: for all three descriptions - `on system in state` -> `one system in state`, `1 formation` -> `one formation`.

File: tests/director/tests/notifications/tenant_hierarchy_notifications_test.go [hunk 2]
```diff
@@ -86,6 +86,10 @@ func TestFormationAssignmentNotificationsTenantHierarchy(stdT *testing.T) {
 
 		subscriptionToken := token.GetClientCredentialsToken(t, ctx, conf.SubscriptionConfig.TokenURL+conf.TokenPath, conf.SubscriptionConfig.ClientID, conf.SubscriptionConfig.ClientSecret, claims.TenantFetcherTenantHierarchyClaimKey)
 		apiPath := fmt.Sprintf("/saas-manager/v1/applications/%s/subscription", conf.SubscriptionProviderAppNameValue)
+
+		subscription.ConfigureTenantTypeAndID(t, httpClient, subscriptionToken, conf.SubscriptionConfig.URL, "customer", "2c4f4a25-ba9a-4dbc-be68-e0beb77a7eb0")
```
Comment (ivantenevvasilev) [inline tests/director/tests/notifications/tenant_hierarchy_notifications_test.go:90]: Maybe a comment as to where this ID is coming from?
`// Customer is Test_DefaultCustomer`?

File: tests/pkg/subscription/utils.go [hunk 3]
```diff
@@ -128,6 +128,57 @@ func CreateSubscription(t *testing.T, conf Config, httpClient *http.Client, appT
 	t.Logf("Successfully created subscription between consumer with subaccount id: %q and tenant id: %q, and provider with name: %q, id: %q and subaccount id: %q", subscriptionConsumerSubaccountID, subscriptionConsumerTenantID, appTmpl.Name, appTmpl.ID, subscriptionProviderSubaccountID)
 }
 
+func ConfigureCostObjectUsage(t *testing.T, httpClient *http.Client, subscriptionToken, subscriptionUrl, parentType, parentID string) {
```
Comment (ivantenevvasilev) [inline tests/pkg/subscription/utils.go:131]: Unused and has the exact same body as `ConfigureTenantTypeAndID`

**Review by ivantenevvasilev — APPROVED** [review]

**Review by ivantenevvasilev — CHANGES_REQUESTED** [review]
seems like a migration has been merged before this one and the timestamp isn't the latest anymore

**Review by ivantenevvasilev — DISMISSED** [review]

**Review by alextargov — APPROVED** [review]

**Review by ivantenevvasilev — APPROVED** [review]

**Review by alextargov — APPROVED** [review]

**Pull request merged**